protocol_driver_grpc_async_callback.cc

```diff
@@ -316,4 +324,178 @@ void ProtocolDriverGrpcAsyncCallback::ShutdownServer() {
   }
 }
 
+namespace {
+class GrpcHandler {
+ public:
+  GrpcHandler(
+      Traffic::AsyncService* service, grpc::ServerCompletionQueue* cq,
+      std::function<std::function<void()>(ServerRpcState* state)>* handler,
+      DistbenchThreadpool* thread_pool)
+      : service_(service),
+        cq_(cq),
+        handler_(handler),
+        responder_(&ctx_),
+        status_(CREATE) {
+    CHECK(thread_pool);
+    thread_pool_ = thread_pool;
+    ProcessRpcFsm();
+  }
+
+  // Unref is called from two places,
+  // one of which will delete the GrpcHandler.
+  void Unref() {
+    if (std::atomic_fetch_sub(&refcnt_, 1) == 1) {
+      delete this;
+    }
+  }
+
+  void ProcessGenericRpc() {
+    rpc_state_.have_dedicated_thread = false;
+    rpc_state_.request = &request_;
+    rpc_state_.SetSendResponseFunction([&]() {
+      response_ = std::move(rpc_state_.response);
+      responder_.Finish(response_, grpc::Status::OK, this);
+    });
+    atomic_fetch_add(&refcnt_, 1);
```

Comment:
similarly third option prevents memory barrier.
It owuld be nice to have a symmetric Ref/Unref instead of inline adjusting the reference count.

Reply:
Agreed